### PR TITLE
Use local post ID in HistoryList to fix TransactionTooLargeException

### LINKED ISSUE #9685: Crash: android.os.TransactionTooLargeException
This one started in 11.8 and has been becoming more prominent in 12.0 and 12.1. 
It's currently the most common crash in Fabric.

Unluckily, the crash logs doesn't say much about where it occurs, but Fabric suggests that it is often seen when transferring bitmaps between Activities, or when saving a large amount of state between Activity configuration changes.

We had other occurrences of it in the past: https://github.com/wordpress-mobile/WordPress-Android/issues/5456.

Also, this thread https://github.com/facebook/react-native/issues/19458 suggests that it may be related to React-Native. The trend we see in Fabric (started in 11.8, growth a bit in 11.9, growth a lot in 12.0 and 12.1) seems to be compatible with mobile Gutenberg deployment in the app.

```
android.os.BinderProxy.transactNative (BinderProxy.java)
android.os.Looper.loop (Looper.java:176)
android.app.ActivityThread.main (ActivityThread.java:6635)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:823)
```

5c93a36df8b88c2963f28894-fabric

### PR DESCRIPTION
**Pull request by shiki**
Fixes #9685 and #5456.

## Findings

When I edit a post with 300K words and put the app to the background, the app will consistently crash.

Using [toolargetool](https://github.com/guardian/toolargetool), I found that the `EditPostActivity`'s `Bundle` grows as the `PostModel` contents grows. At 2000 words:

```
2019-07-16 12:52:01.792 32319-32319/org.wordpress.android D/TooLargeTool: EditPostActivity.onSaveInstanceState wrote: Bundle161877096 contains 16 keys and measures 38.3 KB when serialized as a Parcel
    * stateKeyDroppedMediaUri = 0.1 KB
    * android:viewHierarchyState = 1.0 KB
    * stateKeyRevision = 0.0 KB
    * android:support:fragments = 33.7 KB
	...
```

At 5000 words:

```
2019-07-16 12:52:54.698 32319-32319/org.wordpress.android D/TooLargeTool: EditPostActivity.onSaveInstanceState wrote: Bundle106500119 contains 16 keys and measures 58.7 KB when serialized as a Parcel
    * stateKeyDroppedMediaUri = 0.1 KB
    * android:viewHierarchyState = 1.0 KB
    * stateKeyRevision = 0.0 KB
    * android:support:fragments = 54.1 KB
	...
```

I honed in on the `android:support:fragments` which grows similarly. Digging deeper, I found that `HistoryListFragment` bundles a `key_post` object which is a `PostModel`. At 2000 words:

```
2019-07-16 12:52:02.058 32319-32319/org.wordpress.android D/TooLargeTool: HistoryListFragment.onSaveInstanceState wrote: Bundle27877272 contains 3 keys and measures 2.8 KB when serialized as a Parcel
    * SITE = 1.9 KB
    * android:user_visible_hint = 0.1 KB
    * android:view_state = 0.9 KB
    * fragment arguments = Bundle57837811 contains 2 keys and measures 16.6 KB when serialized as a Parcel
    * key_post = 14.7 KB
    * key_site = 1.9 KB
```

And at 5000 words:

```
2019-07-16 12:52:54.838 32319-32319/org.wordpress.android D/TooLargeTool: HistoryListFragment.onSaveInstanceState wrote: Bundle148956272 contains 3 keys and measures 2.8 KB when serialized as a Parcel
    * SITE = 1.9 KB
    * android:user_visible_hint = 0.1 KB
    * android:view_state = 0.9 KB
    * fragment arguments = Bundle231264648 contains 2 keys and measures 36.9 KB when serialized as a Parcel
    * key_post = 35.1 KB
    * key_site = 1.9 KB
```

If the post contents is even bigger, we would hit the `Bundle` size limit. 

## Solution

This PR changes the `HistoryListFragment` so it will only accept a local post ID and it handles the loading of the `PostModel` itself in the background. When the state is saved, only the local post ID will be saved. The `EditPostActivity` `android:support:fragments` object no longer grows and stays at about `19.1KB`.

Editing the post with 300K words no longer crashes. It takes 5 seconds to load like always but that is currently not the focus of this fix. I tested this with a post that has a lot of images and embedded Youtube videos and I have not observed any crashes.

### Other Changes

Since the `HistoryListFragment` now only has the local post id, I added a simple routine for the unlikely scenario where the `PostModel` cannot be loaded. The history page will show an empty view.

### History List

Please do a regression test for the revision history like:

- Loading the list while online or offline
- Viewing a revision
- Restoring a revision
- Reloading the list when the device is back online
- Swiping to refresh

All the existing features should still work without problems.

### TransactionTooLargeException 

Please also test if you will still receive a `TransactionTooLargeException` crash after this change. There are reproduction steps in the linked issues like:

- https://github.com/wordpress-mobile/WordPress-Android/issues/9685#issuecomment-499279610
- https://github.com/wordpress-mobile/WordPress-Android/issues/9685#issuecomment-493581306

I found that the most consistent way I will receive the crash is:

1. Create a post with a very big content in the Web.
2. On the device, enable Developer → Don't keep activities.
3. Edit the post.
4. Put the app in the background.

The crash happens after this.

## Reviewing

Only 1 reviewer is needed but anyone can review.

## Release Notes

- [x] If there are user-facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/viewmodel/history/HistoryViewModel.kt [hunk 1]
```diff
@@ -107,7 +117,10 @@ class HistoryViewModel @Inject constructor(
 
         revisionAuthorsId = ArrayList(revisionAuthorsId.distinct())
         _revisions.value = getHistoryListItemsFromRevisionModels(revisions)
-        fetchRevisionAuthorDetails(revisionAuthorsId)
+
+        if (revisionAuthorsId.isNotEmpty()) {
```
Comment (shiki) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/history/HistoryViewModel.kt]: I had to add this check in order to make this part of the class testable. We're currently only testing for empty `revisionAuthorsId`. If it's not empty, the `fetchRevisionAuthorDetails` will crash in unit tests.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/HistoryListFragment.kt [hunk 2]
```diff
@@ -151,5 +149,13 @@ class HistoryListFragment : Fragment() {
                 (activity as HistoryItemClickInterface).onHistoryItemClicked(it, viewModel.revisionsList)
             }
         })
+
+        viewModel.post.observe(this, Observer { post ->
+            actionable_empty_view.subtitle.text = if (post?.isPage == true) {
```
[the remark this answers is not in the document]
Comment (shiki) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/HistoryListFragment.kt]: I think it's fine. The original code assumed that `post` will never be `null`:

```kt
        actionable_empty_view.subtitle.text = if ((arguments?.get(KEY_POST) as PostModel).isPage) {
```

With this change, it _can_ be but unlikely. If it happens that it's `null`, I don't think it's terrible that we show the `R.string.history_empty_subtitle_post` message.

**Comment by shiki** [comment]
@malinajirka Thank you for spending the time to test this! 

> However, I wasn't able to reproduce the crash on any of my devices/emulators. The EditPostActivity either never loads (black screen or frozen editor screen) or it doesn't crash.

That's weird. And no errors in the logs too? 

> Having said that, I've encountered the TransactionTooLargeException in many other applications and I'm positive this fix will solve it

Thanks for confirming that. This is the first time I've seen and fixed an error like this. 😄 

> We still store SiteModel into a Bundle at a lot of places, but it shouldn't be that big hopefully

Right. I looked at them and they weren't that big. Is it worth it to create an issue for it? 

I responded to your concern. Please let me know if I merge this. 🙂

**Comment by malinajirka** [comment]
>That's weird. And no errors in the logs too?

Yep, no errors. I have encountered some segmentation fault in C code once, but the log wasn't very helpful what happened and I wasn't able to reproduce it.

>Right. I looked at them and they weren't that big. Is it worth it to create an issue for it?

I don't think it's necessary to fix them. We just need to make sure we don't introduce the same issue in any new code. I know we have discussed this issue about a year ago, so hopefully at least most people should be aware that it's not a best practice to store potentially huge objects into Bundles. Thank you though;)!

Looks good to me. Thank you, great job!

**Pull request merged**